### PR TITLE
vmware,storage: fix solidfire ROOT volume creation with vmware 6.5 and above

### LINKED ISSUE #3598: SOLIDFIRE: "SolidFire" plugin doesn't work for ROOT volumes with VMware (6.5)
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
SolidFire plugin ('SolidFire" as opposed to "SolidFire Shared")
~~~

##### CLOUDSTACK VERSION

~~~
4.13 (master atm), but also observed by another community member on 4.11.3
~~~

##### OS / ENVIRONMENT

VMware 6.5 tested

##### SUMMARY

Adding SolidFire Primary Storage via SolidFire plugin for VMware 6.5 fails with errors in mounting Datastore in ESXi hosts and raises the specific error in mgmt logs.

##### STEPS TO REPRODUCE

in vCenter, add iSCSI Software adapter  to each ESXi hosts, configure proper network binding to vSwitchXXX for the iSCSI traffic (i.e. VLAN XXX so that ESXi can communicate with the SVIP on that VLAN), then add SolidFire (Zone wide (same problem with cluster-wide), protocol "Custom, provider "SolidFire", "Managed" box ticked and proper ULR - adding SF as Primary Storage is successful.

Try to spin a VM - that is when the things fail after a minutes or so.

Observing vCenter, the following thing happen (also check the screenshot)

- Static iSCSI target is added to ESXI hosts
- Rescanning HBAs
- Creating datastore same size as the volume/template itself with the name ending in "Centos-5.3-x64" or similar (name of the template)
- Deploying OVF template
- Unregistering VM
- Moving files around
- unmounting VMFS
- Removing iSCSI static targets
- Rescan HBA
- Again adding iSCSI static targets **???**
- Rescan HBAs
- Rescan VMFS
- RENAME datastore (from the template-alike name to the  root-volume-alike name, ending with ROOT-XXX.YY)
**(... probably NOW problem happens ???)))**
- unmount datastore
- remove iSCSI targets.

The error from the ACS is:
_message: Datastore '-iqn.2010-01.com.solidfire:hl1k.root-32.29-0' is not accessible. No connected and accessible host is attached to this datastore_

The problem is - this datastore (in it's latest, renamed state, still attached to ESXi hosts) - is unmounted, but can't be removed, NOR can I mount it - If I try to manually mount it, I get the vCenter message of "Operation failed, diagnostics report: Unable to find volume uuid[5d7abd9a-273aa9d5-bffe-1e00d4010711] lvm [snap-329aa3ea-5d7abd01-a5c83210-c87c-1e00d4010711] devices"

Screenshot from vCenter attached - note that the last 2 entries (on top) are my attempt to manually mount an existing SF datastore. - i.e. there are zero failures on vCenter side while ACS is doing it's job - something is failing on ACS side.

image

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #3598 

Continue #4977 

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java [hunk 1]
```diff
@@ -314,16 +316,19 @@ public String[] listDirContent(String path) throws Exception {
...
-        HostDatastoreBrowserMO browserMo = getHostDatastoreBrowserMO();
-
-        s_logger.info("Search file " + file.getFileName() + " on " + dirFile.getPath());
-        HostDatastoreBrowserSearchResults results = browserMo.searchDatastore(dirFile.getPath(), file.getFileName(), true);
-        if (results != null) {
-            List<FileInfo> info = results.getFile();
-            if (info != null && info.size() > 0) {
-                s_logger.info("File " + fileFullPath + " exists on datastore");
-                return true;
+        Boolean folderExists = true;
```
Comment (sureshanaparti) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java:319]: is this true by default?

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java [hunk 2]
```diff
@@ -314,16 +316,19 @@ public String[] listDirContent(String path) throws Exception {
...
+        if(StringUtils.isNotBlank(file.getDir())){
```
Comment (sureshanaparti) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java:320]:
```suggestion
        if(StringUtils.isNotBlank(file.getDir())) {
```

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java [hunk 3]
```diff
@@ -314,16 +316,19 @@ public String[] listDirContent(String path) throws Exception {
...
+            folderExists = folderExists(String.format("[%s]", file.getDatastoreName()), file.getDir());
+        }
+        if (folderExists){
```
Comment (sureshanaparti) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/DatastoreMO.java:323]:
```suggestion
        if (folderExists) {
```

File: plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java [hunk 4]
```diff
@@ -3117,6 +3117,9 @@ private void removeVmfsDatastore(Command cmd, VmwareHypervisorHost hyperHost, St
 
     private void removeVmfsDatastore(Command cmd, VmwareHypervisorHost hyperHost, String datastoreName, String storageIpAddress, int storagePortNumber,
                                      String iqn, List<Pair<ManagedObjectReference, String>> lstHosts) throws Exception {
+        if (HypervisorHostHelper.isHostVersionEqualOrHigher(hyperHost, "6.5")) {
```
Comment (sureshanaparti) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/storage/resource/VmwareStorageProcessor.java:3120]: does this consider minor release version as well?

**Comment by tsinik-dw** [comment]
Hi all, 

I have some news regarding this PR.

I just did some tests using ACS 4.16.1, Solidfire Demo Node with Element OS v.12.5.0.897 and VMware 6.7
Surprisingly, I managed to create VMs with ROOT volumes on the Solidfire.
I even managed to perform volume migration between primary storages (from Solidfire to NFS and from NFS back to Solidfire).
If you need any logs or other info I am willing to help.

My last failed attempt was using ACS 4.16.0, Solidfire Demo Node with Element OS v.12.3.x.x and VMware 6.7
I don't know what changed but the VMs were created on Solidfire without any complaints. 
Can anyone else confirm this?

**Comment by rohityadavcloud** [comment]
cc @harikrishna-patnala was any work done towards general storage work that could have fixed the issues with solidfire (as a side effect)?

**Comment by shwstppr** [comment]
Based on the comments ^^, I'm closing this for now. Will re-open if later needed. cc @rohityadavcloud @pdion891

**Comment by weizhouapache** [comment]
> Hi all,
> 
> I have some news regarding this PR.
> 
> I just did some tests using ACS 4.16.1, Solidfire Demo Node with Element OS v.12.5.0.897 and VMware 6.7 Surprisingly, I managed to create VMs with ROOT volumes on the Solidfire. I even managed to perform volume migration between primary storages (from Solidfire to NFS and from NFS back to Solidfire). If you need any logs or other info I am willing to help.
> 
> My last failed attempt was using ACS 4.16.0, Solidfire Demo Node with Element OS v.12.3.x.x and VMware 6.7 I don't know what changed but the VMs were created on Solidfire without any complaints. Can anyone else confirm this?

@tsinik-dw 
thanks for your update.

Do you use cloudstack with solidfire storage in your production ? have you faced any issues ?

**Pull request closed without merging**